pkg/sdkserver/sdkserver.go

```diff
@@ -762,20 +790,202 @@ func (s *SDKServer) GetCounter(ctx context.Context, in *alpha.GetCounterRequest)
+		}
+		capacitySet := in.CounterUpdateRequest.Capacity.GetValue()
+		batchCounter.capacitySet = &capacitySet
+	default:
+		return nil, errors.Errorf("INVALID_ARGUMENT. Malformed CounterUpdateRequest: %v", in.CounterUpdateRequest)
+	}
+
+	s.gsCounterUpdates[name] = batchCounter
+
+	// Queue up the Update for later batch processing by updateCounters.
+	s.workerqueue.Enqueue(cache.ExplicitKey(updateCounters))
+	return &alpha.Counter{}, nil
+}
+
+// updateCounter updates the Counters in the GameServer's Status with the batched update requests.
+func (s *SDKServer) updateCounter(ctx context.Context) error {
+	if !runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
+		return errors.Errorf("%s not enabled", runtime.FeatureCountsAndLists)
+	}
+
+	gs, err := s.gameServer()
+	if err != nil {
+		return err
+	}
+	gsCopy := gs.DeepCopy()
+
+	s.logger.WithField("batchCounterUpdates", s.gsCounterUpdates).Debug("Batch updating Counter(s)")
+	s.gsUpdateMutex.Lock()
+	defer s.gsUpdateMutex.Unlock()
+
+	for name, ctrReq := range s.gsCounterUpdates {
+		counter, ok := gsCopy.Status.Counters[name]
+		if !ok {
+			continue
+		}
+		// If the Counter is the same as the original Counter stored in the counterUpdateRequest
+		// then no other changes have been made to this Counter, and we can overwrite the values.
+		if cmp.Equal(counter, ctrReq.counter) {
+			count := counter.Count
+			if ctrReq.countSet != nil {
+				count = *ctrReq.countSet
+			}
+			count += ctrReq.diff
+			counter.Count = count
+			if ctrReq.capacitySet != nil {
+				counter.Capacity = *ctrReq.capacitySet
+			}
+		} else {
+			// If the Counter is NOT the same as the original Counter stored in the counterUpdateRequest
+			// then changes have been made to the Counter since we validated the incoming changes in
+			// UpdateCounter, and we need to verify if the batched changes can be fully applied, partially
+			// applied, or cannot be applied.
+			if ctrReq.capacitySet != nil {
+				counter.Capacity = *ctrReq.capacitySet
+			}
+			if ctrReq.countSet != nil {
+				counter.Count = *ctrReq.countSet
+			}
+			newCnt := counter.Count + ctrReq.diff
+			if newCnt < 0 {
+				// If Count < 0 truncate requests by setting the Count to 0.
+				newCnt = 0
+				s.logger.Debug("truncating Count in Update Counter request to 0")
+			}
+			// TODO: This means there may be increment requests in the collapsed Diff that shouldn't have gone through. We may want a diffList for this case?
+			// If Count > Capacity and the Diff is decreasing the Count, reduce the Count by the Diff.
+			// Otherwise if Count > Capacity and the Diff is increasing the Count, truncate at Capacity.
+			if newCnt > counter.Capacity && ctrReq.diff > 0 {
+				newCnt = counter.Capacity
+				s.logger.Debug("truncating Count in Update Counter request to Capacity")
+			}
+			counter.Count = newCnt
+		}
+		gsCopy.Status.Counters[name] = counter
+		gs, err = s.gameServerGetter.GameServers(s.namespace).Update(ctx, gsCopy, metav1.UpdateOptions{})
```

Comment:
If I'm reading this right, are we doing an update per count key that is batched? 

If so, we should instead apply all batch requests into a single update - we shouldn't be doing _n_ updates, where _n_ is the number of keys waiting to be batched.

I think we can still do multiple events after the fact, but we should consolidate updates into a single `s.gameServerGetter.GameServers(s.namespace).Update(ctx, gsCopy, metav1.UpdateOptions{})` operation.

(Unless I'm reading the formatting of this wrong).

Then we can clear the entire `s.gsCounterUpdates` on completion, which should also mean you can remove the entire `if cmp.Equal(counter, ctrReq.counter)` section, since the key will either be there or not.

Reply:
> I think we can still do multiple events after the fact

What's the downside to do only one event for the single consolidated update?

Reply:
It's more of a readability issue when looking at the event stream. If we have 5 keys in a batch, it's probably easier to read an event per key after the fact, than try and smoosh all the info from 5 keys into a single event.

(To also be clear, I was thinking one event per key).

Reply:
Yes, there's an update for each key (each Counter), not an update for the whole Counters map, the disadvantage of that being obvious. The advantage is that you get an accurate per-Counter updating and event log, and updating one Counter doesn't depend on another.

I can move to updating the whole Counters map at once instead of per Counter. I would just need to make sure the event recording still happened after the update even if it's per key, so I'd probably pull that out into a separate method.

Reply:
If you do one Update for all counters - as long as you apply the Events after the fact, it's pretty much fire and forget, so I don't expect it to be an issue.